File: SistemaEletronico.js
class SistemaAutenticacao {
  autenticarUsuario(usuario, senha) {
    console.log(`Autenticacao do usuario ${usuario}....`);

    if (usuario === "maria@gmail.com" && senha === "maria24") {
      console.log("Acesso Permitido");
      console.log("-----------------------");
      return true;
    } else if (usuario !== "" && senha !== "") {
      console.log("Acesso Negado");
      console.log("-----------------------");
      return false;
    }
  }
}

class SistemaCadastro {
  cadastrarUsuario(usuario) {
    console.log("-----------------------");
    console.log(`Cadastrando o usuario ${usuario}....`);
    console.log(`Bem Vindo(a)...`);
    console.log("-----------------------");
  }
}

class SistemaPedido {
  criarPedido(pedido) {
    console.log(`Criando o pedido ${pedido}....`);
  }
}

class AtendimentoCliente {
  gerarID() {
    return Math.random().toString(36).substr(2, 9);
    //Cria um ID com 9 letras misturado com numeros
  }

  realizarAtendimento(usuario, senha, produto) {
    const autenticacao = new SistemaAutenticacao();
    const cadastro = new SistemaCadastro();
    const pedido = new SistemaPedido();

    if (autenticacao.autenticarUsuario(usuario, senha)) {
      const pedidoID = this.gerarID();
      const ID = pedidoID;
      pedido.criarPedido(produto, usuario);
      cadastro.cadastrarUsuario(usuario);
      console.log(`Atendimento realizado com sucesso. Pedido ID : ${pedidoID}`);
    } else {
      console.log(`Falha na autenticação. Atendimento não realizado`);
    }
    return "Camisa";
  }
}
//Criacao do Proxy
class AtendimentoClienteProxy {
  constructor() {
    this.atendimentoCliente = new AtendimentoCliente();
  }
  //realizarAtendimento irá chamar outra funcao do msm nome acima
  realizarAtendimento(usuario, senha, produto) {
    console.log("Interceptando requisição...");
    this.atendimentoCliente.realizarAtendimento(usuario, senha, produto);
  }
}

//Uso do Proxy
const proxy = new AtendimentoClienteProxy();
// E-mail ou senha diferente, irá dar erro
const usuario = "maria@gmail.com";
const senha = "maria24";
const produto = "Camisa";
proxy.realizarAtendimento(usuario, senha, produto);

File: SistemaEletronico.py
import random

# Subsistema 1
class SistemaAutenticacao:
    def autenticar_usuario(self, usuario, senha):
        if usuario == "fulano@gmail.com" and senha == "usuario123":
            print("Acesso Permitido")
            return True
        else:
            print("Acesso Negado")
            return False

# Proxy para o Subsistema 1
class SistemaAutenticacaoProxy:
    def __init__(self):
        self.sistema_autenticacao = SistemaAutenticacao()
        self.usuario_autenticado = False

    def autenticar_usuario(self, usuario, senha):
        if self.usuario_autenticado:
            print("Usuário já está autenticado.")
            return True
        else:
            return self.sistema_autenticacao.autenticar_usuario(usuario, senha)

# Subsistema 2
class SistemaCadastro:
    def __init__(self):
        self.usuarios_cadastrados = []

    def cadastrar_usuario(self, usuario):
        print(f"Cadastrando usuário {usuario}...")
        self.usuarios_cadastrados.append(usuario)
        return True  # Simulado deu certo

# Subsistema 3
class SistemaPedido:
    def criar_pedido(self, produtos):
        pedido_id = f"PED-{random.randint(100, 999)}"  # ID aleatório de 3 números
        print(f"Criando pedido de produtos {produtos} com ID: {pedido_id}")
        return pedido_id

# Fachada (Facade)
class AtendimentoCliente:
    def __init__(self):
        self.autenticacao_proxy = SistemaAutenticacaoProxy()
        self.cadastro = SistemaCadastro()
        self.pedido = SistemaPedido()

    def realizar_atendimento(self, usuario, senha, produtos):
        if self.autenticacao_proxy.autenticar_usuario(usuario, senha):
            pedido_id = self.pedido.criar_pedido(produtos)
            self.cadastro.cadastrar_usuario(usuario)
            print(f"Atendimento realizado com sucesso. Pedido: {pedido_id}")
        else:
            print("Falha na autenticação. Atendimento não realizado.")

# Cliente
atendimento = AtendimentoCliente()

# usuario = "shaolin123@gmail.com" # // Para testar possíveis falhas na autenticação
# senha = "shaolin1234"

usuario = "fulano@gmail.com"
senha = "usuario123"
produtos = ["Fita RGB", "Caneta Esferográfica", "Empire: Total War"]

atendimento.realizar_atendimento(usuario, senha, produtos)
